internal/transformers/estransforms/optionalchain.go

```diff
@@ -11,7 +11,222 @@ type optionalChainTransformer struct {
+		capturedLeft = left.AsSyntheticReferenceExpression().Expression
+	}
+	leftExpression := ch.Factory().RestoreOuterExpressions(expression, capturedLeft, ast.OEKPartiallyEmittedExpressions)
+	if !transformers.IsSimpleCopiableExpression(capturedLeft) {
+		capturedLeft = ch.Factory().NewTempVariable()
+		ch.EmitContext().AddVariableDeclaration(capturedLeft)
+		leftExpression = ch.Factory().NewAssignmentExpression(capturedLeft, leftExpression)
+	}
+	rightExpression := capturedLeft
+	var thisArg *ast.Expression
+
+	for i, segment := range chain {
+		switch segment.Kind {
+		case ast.KindElementAccessExpression, ast.KindPropertyAccessExpression:
+			if i == len(chain)-1 && captureThisArg {
+				if !transformers.IsSimpleCopiableExpression(rightExpression) {
+					thisArg = ch.Factory().NewTempVariable()
+					ch.EmitContext().AddVariableDeclaration(thisArg)
+					rightExpression = ch.Factory().NewAssignmentExpression(thisArg, rightExpression)
+				} else {
+					thisArg = rightExpression
+				}
+			}
+			if segment.Kind == ast.KindElementAccessExpression {
+				rightExpression = ch.Factory().NewElementAccessExpression(rightExpression, nil, ch.Visitor().VisitNode(segment.AsElementAccessExpression().ArgumentExpression), ast.NodeFlagsNone)
+			} else {
+				rightExpression = ch.Factory().NewPropertyAccessExpression(rightExpression, nil, ch.Visitor().VisitNode(segment.AsPropertyAccessExpression().Name()), ast.NodeFlagsNone)
+			}
+		case ast.KindCallExpression:
+			if i == 0 && leftThisArg != nil {
+				if !ch.EmitContext().HasAutoGenerateInfo(leftThisArg) {
+					leftThisArg = leftThisArg.Clone(ch.Factory())
+					ch.EmitContext().AddEmitFlags(leftThisArg, printer.EFNoComments)
+				}
+				callThisArg := leftThisArg
+				if leftThisArg.Kind == ast.KindSuperKeyword {
+					callThisArg = ch.Factory().NewThisExpression()
+				}
+				rightExpression = ch.Factory().NewFunctionCallCall(rightExpression, callThisArg, ch.Visitor().VisitNodes(segment.ArgumentList()).Nodes)
+			} else {
+				rightExpression = ch.Factory().NewCallExpression(
+					rightExpression,
+					nil,
+					nil,
+					ch.Visitor().VisitNodes(segment.ArgumentList()),
+					ast.NodeFlagsNone,
+				)
+			}
+		}
+		ch.EmitContext().SetOriginal(rightExpression, segment)
+	}
+
+	var target *ast.Node
+	if isDelete {
+		target = ch.Factory().NewConditionalExpression(
+			createNotNullCondition(ch.EmitContext(), leftExpression, capturedLeft, true),
+			ch.Factory().NewToken(ast.KindQuestionToken),
+			ch.Factory().NewTrueExpression(),
+			ch.Factory().NewToken(ast.KindColonToken),
+			ch.Factory().NewDeleteExpression(rightExpression),
+		)
+	} else {
+		target = ch.Factory().NewConditionalExpression(
+			createNotNullCondition(ch.EmitContext(), leftExpression, capturedLeft, true),
+			ch.Factory().NewToken(ast.KindQuestionToken),
+			ch.Factory().NewVoidZeroExpression(),
+			ch.Factory().NewToken(ast.KindColonToken),
+			rightExpression,
+		)
+	}
+	target.Loc = node.Loc
```

Comment:
Any reason you have to do this? I see you run `SetOriginal` farther below. The ordering is a little bit different here though. I would have thought you'd run `SetOriginal` here based on the original code.